libconnman-qt/useragent.cpp

```diff
@@ -12,39 +12,65 @@
 
 static const char AGENT_PATH[] = "/ConnectivityUserAgent";
 
-UserAgent::UserAgent(QObject* parent) :
-    QObject(parent),
-    m_req_data(nullptr),
-    m_manager(NetworkManager::sharedInstance()),
-    requestType(TYPE_DEFAULT),
-    agentPath(QString())
+class UserAgentPrivate
+{
+public:
+    enum ConnectionRequestType {
+        TYPE_DEFAULT =0,
+        TYPE_SUPPRESS,
+        TYPE_CLEAR
+    };
+
+    UserAgentPrivate();
+
+    ServiceRequestData *m_req_data;
+    QSharedPointer<NetworkManager> m_manager;
+    QDBusMessage currentDbusMessage;
+    ConnectionRequestType requestType;
+    QString agentPath;
+    QTimer requestTimer;
+    QDBusMessage requestMessage;
+};
+
+UserAgentPrivate::UserAgentPrivate()
+    : m_req_data(nullptr)
+    , m_manager(NetworkManager::sharedInstance())
+    , requestType(TYPE_DEFAULT)
+{
+}
+
+UserAgent::UserAgent(QObject* parent)
+    : QObject(parent)
+    , d_ptr(new UserAgentPrivate)
 {
     QString agentpath = QLatin1String("/ConnectivityUserAgent");
     setAgentPath(agentpath);
-    connect(m_manager.data(), &NetworkManager::availabilityChanged,
+    connect(d_ptr->m_manager.data(), &NetworkManager::availabilityChanged,
             this, &UserAgent::updateMgrAvailability);
 
-    requestTimer = new QTimer(this);
-    requestTimer->setSingleShot(true);
-    connect(requestTimer, &QTimer::timeout,
+    d_ptr->requestTimer.setSingleShot(true);
+    connect(&d_ptr->requestTimer, &QTimer::timeout,
             this, &UserAgent::requestTimeout);
 }
 
 UserAgent::~UserAgent()
 {
-    m_manager->unregisterAgent(QString(agentPath));
+    d_ptr->m_manager->unregisterAgent(QString(d_ptr->agentPath));
+
+    delete d_ptr;
+    d_ptr = nullptr;
```

Comment:
possible d_ptr->m_req_data leak? should call cancelUserInput() first or something?

Reply:
Looks like a real deal. I'll add a commit for this.